src/net/poll.rs

```diff
@@ -169,18 +169,10 @@ impl NetworkState {
     }
 
     /// Connect to a remote peer, but don't register it with the poll handle.
-    /// Connect *synchronously*, and then put the socket into non-blocking mode.
+    /// The underlying connect(2) is _asynchronous_, so the caller will need to register it with a
+    /// poll handle and wait for it to be connected.
     pub fn connect(&mut self, addr: &SocketAddr) -> Result<mio_net::TcpStream, net_error> {
-        // connect and abort after 5 seconds of waiting.
-        // TODO: revert to non-blocking connect to avoid slow neighbors starving socket work
-        let inner_stream = net::TcpStream::connect_timeout(addr, time::Duration::from_millis(5000))
-            .map_err(|_e| {
-                test_debug!("Failed to connect to {:?}: {:?}", addr, &_e);
-                net_error::ConnectionError
-            })?;
-
-        // NOTE: this will put the socket into non-blocking mode
-        let stream = mio_net::TcpStream::from_stream(inner_stream)
+        let stream = mio_net::TcpStream::connect(addr)
```

Comment:
Is this going to cause the same issue with non-determinism in tests? Maybe add a `sync: bool` parameter?

Reply:
It shouldn't.  First, only the `PeerNetwork` and `HttpPeer` structs call this method (and they do so internally, as part of initiating a new conversation).  Second, both of these structs now keep the socket in a "pending connection" state, so no messages will be sent on the socket until the underlying `connect(2)` completes.  See the new code in `process_ready_sockets()` in both structs -- there's now a `self.connecting` map that holds sockets that have not yet finished connecting, which gets updated once the poller detects a writeable event for the socket.